### PR TITLE
bpo-31819: Add AbstractEventLoop.sock_recv_into()

### PR DESCRIPTION
**Pull request by pitrou**
https://bugs.python.org/issue31819

### CODE REVIEW

File: Lib/asyncio/selector_events.py [hunk 1]
```diff
@@ -386,6 +386,41 @@ def _sock_recv(self, fut, registered, sock, n):
...
+        The received data is written into *buf* (a writable buffer).
+        The return value is the number of bytes written.
+
+        This method is a coroutine.
+        """
+        if self._debug and sock.gettimeout() != 0:
+            raise ValueError("the socket must be non-blocking")
+        fut = self.create_future()
+        self._sock_recv_into(fut, False, sock, buf)
+        return fut
```
Comment (1st1) [inline Lib/asyncio/selector_events.py:401]: If you want, you can further optimize this by transforming `sock_recv_into` into an `async def` coroutine, and creating the `fut` object only when `sock.recv_into(buf)` fails with a `BlockingError`.

Comment (pitrou) [reply Lib/asyncio/selector_events.py:401]: I don't think that's very useful, as people will mainly want to use that for large data. Right now the implementation mirrors that of `sock_recv`, which is easier to understand.

**Comment by 1st1** [comment]
@pitrou FWIW I've added `loop.sock_recv_into()` to uvloop.  The next version with this change will be released next week.

**Comment by pitrou** [comment]
Great! Thank you @1st1.

File: Lib/asyncio/windows_events.py [hunk 2]
```diff
@@ -448,6 +448,28 @@ def finish_recv(trans, key, ov):
...
+    def recv_into(self, conn, buf, flags=0):
+        self._register_with_iocp(conn)
+        ov = _overlapped.Overlapped(NULL)
+        try:
+            if isinstance(conn, socket.socket):
+                ov.WSARecvInto(conn.fileno(), buf, flags)
+            else:
+                ov.ReadFileInto(conn.fileno(), buf)
+        except BrokenPipeError:
+            return self._result(b'')
```
Comment (vstinner) [inline Lib/asyncio/windows_events.py:460]: The doc says "The return value is the number of bytes written." but here an empty byte string is returned! I created https://bugs.python.org/issue41467 to track this bug.

**Pull request merged**